src/mlpack/core/optimizers/Nadam/nadam.hpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file nadam.hpp
+ * @author Sourabh Varshney
+ *
+ * Nadam optimizer. Nadam is an optimizer that combines the effect of Adam and
+ * NAG to the gradient descent to improve its Performance.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_NADAM_NADAM_HPP
+#define MLPACK_CORE_OPTIMIZERS_NADAM_NADAM_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include <mlpack/core/optimizers/sgd/sgd.hpp>
+#include "nadam_update.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * Nadam is an optimizer that combines the Adam and NAG.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @article{
+ *   author  = {Sebastian Ruder},
+ *   title   = {An overview of gradient descent optimization algorithms},
+ *   journal = {CoRR},
+ *   year    = {2016},
+ *   url     = {https://arxiv.org/abs/1609.04747v2}
+ * }
+ * @endcode
+ *
+ *
+ * For Nadam to work, a DecomposableFunctionType template parameter
+ * is required. This class must implement the following function:
+ *
+ *   size_t NumFunctions();
+ *   double Evaluate(const arma::mat& coordinates,
+ *                   const size_t i,
+ *                   const size_t batchSize);
+ *   void Gradient(const arma::mat& coordinates,
+ *                 const size_t i,
+ *                 arma::mat& gradient,
+ *                 const size_t batchSize);
+ *
+ * NumFunctions() should return the number of functions (\f$n\f$), and in the
+ * other two functions, the parameter i refers to which individual function (or
+ * gradient) is being evaluated.  So, for the case of a data-dependent 
+ * function, such as NCA (see mlpack::nca::NCA), NumFunctions() should return 
+ * the number of points in the dataset, and Evaluate(coordinates, 0) will 
+ * evaluate the objective function on the first point in the dataset 
+ * (presumably, the dataset is held internally in the 
+ * DecomposableFunctionType).
+ *
+ * @tparam UpdateRule Nadam optimizer update rule to be used.
+ */
+template<typename UpdateRule = NadamUpdate>
+class NadamType
```

Comment:
I think if we follow, the paper: https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ there is no need to introduce a new ``NadamType`` class, we could just use the existing ``AdamType`` class and just add the ``NadamRule``. Let me know what you think.

Reply:
@zoq I want to know how the decay parameter is updated?

Reply:
Yes the same class can be used. Just we need to use the Nadam Update Rule. I would do it.